docs/blog/2019-06-10-building-a-jobs-board-a-case-study/index.md

```diff
@@ -0,0 +1,40 @@
+
+Instead we need to build our job site with a frontend client that queries API (NodeJS) endpoints at runtime, the API will then query a database (MongoDB, for flexibility) to create or list a job. With this setup we can, post a job, and asynchronously update the main front page to include the latest listing when the transaction has been processed. 
+
+Now that we have a system design we are happy with, we can talk the false starts that inspired this current design since you know, hindsight is 20/20.
+
+# Gotchas and false starts along the way
+
+The biggest gotcha focused on job listing SEO and the server-side rendering of job detail pages. The detail page was designed to fetch data for a specific job post based on the unique ID passed through from the front page. We implemented this feature using client-only routes because we figured this approach to be the "Gatsby way" of working with apps, based on an article focused on [building hybrid apps](https://www.gatsbyjs.org/docs/building-apps-with-gatsby/) with Gatsby.
+
+However we found out that these pages could not use React Helmet to add meta tags for social media sharing. The culprit of this problem lies in the fact that Gatsby pages are served using the React.Hydrate function which is how Gatsby generates server rendered pages, processed with ExpressJS at build time. Pages rendered via React.Hydrate load twice on the frontend. On first call, the React lifecycle methods are unavailable afterwards, lifecycle methods are available to make API calls. Due to this rendering process we were not able to make API calls in `componentDidMount` to have the data present before we added SEO tags with React Helmet. Boo!
+
+To solve this problem we added request forwarding at the CDN level and moved the details rendering logic to our backend ExpressJS app. When the user requests a job post details, the request goes to a backend app, we generate a job post with the proper SEO meta tags, return a response with the proper data, and then cache the response for future requests. After we made these adjustments we were able to have shareable job posts. Yay!
+
+# Final thoughts
+
+To recap, we built a remote jobs board with Gatsby as our frontend client because we wanted performance from the start. We deployed our site on [Netlify](https://netlify.com), configured CDN request forwarding with Netlify, used a backend ExpressJS app, deployed on [Google Cloud Appengine](https://cloud.google.com/appengine/), to handle the serving of job detail pages. We hit a frustrating blocker for a critcial feature that caused us to discard core Gatsby logic and reimplement that logic on the backend. Given what we know now, would we choose Gatsby again? Absolutely! Though we were not able to leverage Gatsby to its full potential, we're confident that many more opportunities are on the horizon. To learn more about me you can follow me [@stevennatera](https://twitter.com/stevennatera) on Twitter. Thanks!
```

Comment:
what did you discard? Is there an issue around this?

Reply:
oh the hydrate thing — not really sure what you mean by double-render? Is this what Gatsby normally does or a side-effect of your expressjs processing?

Reply:
Heya Kyle. Sorry to pop in unannounced here, and please let me know if there's a more appropriate place for this discussion, but today my team discovered the issue that nodox is referring to, specifically, that we aren't able to populate our meta title/description for routes corresponding to user-generated content, since the page component has to make an API call, to determine what metadata to inject with Helmet - which happens after the initial render / Helmet metadata stamping. This is such an issue for is that we're debating whether we need to move off Gatsby. (Which we otherwise love!) Is there any recommendations you might be aware of around this?

Reply:
@Elementrat that seems like a relatively known issue -- you're saying that dynamic content doesn't get the same SEO niceties as build-time (fully SSR content) right?

In the future as we roll out offerings like Incremental Builds, this will be less of an issue (or a non-issue!) because it'll be cheaper/easier/etc. to fire off a build, re-build only what changed (e.g. 3-4 pages or something), and do so very quickly (in a few minutes or less).

Reply:
What @Elementrat pointed to was exactly what I was getting at. Would like for their to be a solution for this in a way where rebuilding the site is cheap and fast. @DSchau sounds like a promising solution.

Most of the Hydrate functionality is well documented but I was unsure I was going to fix this when I first encountered it with gatsby. I looked at my code again and I actually kept the client-only routes but supplemented that with a Nodejs endpoint that would handle direct requests via the CDN request forwarding.